lib/elixir/pages/anti-patterns/macro-anti-patterns.md

```diff
@@ -107,6 +107,115 @@ iex> MyMath.sum(3+1, 5+6)
+need to be recompiled.
+
+This becomes especially problematic when projects grow and the number of compile-connected
+dependencies is not kept under control, making modifying any file trigger a recompilation
+of many other files and slowing down the development process.
+
+Note that depending on modules from external libraries is not a problem, as these modules
+are only compiled once when the library is compiled.
+
+#### Example
+
+The code below defines three modules `ModuleA`, `Macros` and `MacroUtils`, each in their
+own file. The dependencies between the three files are as follows:
+
+- `lib/module_a.ex` has a _compile_ dependency on `lib/macros.ex` because `ModuleA`
+  calls a macro defined in `Macros`,
+- `lib/macros.ex` has a _runtime_ dependency on `lib/macro_utils.ex` because `Macros`
+  calls a function defined in `MacroUtils` within the body of its macro.
+
+In this example, `lib/macro_utils.ex` is a compile dependency of `lib/module_a.ex`,
+through a transitive dependency via `lib/macros.ex`. This means that modifying
+`lib/macro_utils.ex` will trigger a recompilation of `lib/module_a.ex`, as if
+`lib/module_a.ex` had a direct compile dependency on `lib/macro_utils.ex`.
+
+Note that if `MacroUtils` itself depended on other modules directly or indirectly,
+modifying the files containing these modules would also trigger a recompilation of
+`lib/module_a.ex`.
+
+```elixir
+# lib/module_a.ex
+defmodule ModuleA do
+  require Macros
+
+  def hello do
+    Macros.greet()
+  end
+end
+```
+
+```elixir
+# lib/macros.ex
+defmodule Macros do
+  defmacro greet do
+    module = MacroUtils.get_module(__CALLER__)
+
+    quote do
+      IO.puts("Hello from #{unquote(module)}")
+    end
+  end
+end
+```
+
+```elixir
+# lib/macro_utils.ex
+defmodule MacroUtils do
+  def get_module(caller) do
+    caller.module
+  end
+end
+```
+
+#### Refactoring
+
+To remove this pattern, the developer must remove any dependency to other modules from
+the same project. In the code shown below, the `MacroUtils` module was deleted and the
+`get_module/1` function was moved to the `Macros` module. This way, the `Macros` module
+no longer depends on any other module.
+
+```elixir
+# lib/macros.ex
+defmodule Macros do
+  defmacro greet do
+    module = get_module(__CALLER__)
+
+    quote do
+      IO.puts("Hello from #{unquote(module)}")
+    end
+  end
+
+  defp get_module(caller) do
+    caller.module
+  end
+end
+```
+
+#### Additional remarks
+
+Whilst this guide only covers macros, modules can be compile dependencies of other modules
+in other ways. For example, calling another module in the module, in a module attribute
+for example, will also create a compile dependency. Some external libraries may also
+create compile dependencies by using macros.
+[`mix xref`](https://hexdocs.pm/mix/main/Mix.Tasks.Xref.html#module-a-brief-introduction-to-xref)
+can be useful to identify these dependencies and refactor the code to remove them.
```

Comment:
I just pushed an anti-pattern on compile-time dependencies, which I think is the most appropriate for this guide, inspired by the conversation started here: https://github.com/elixir-lang/elixir/blob/main/lib/elixir/pages/anti-patterns/macro-anti-patterns.md#compile-time-dependencies

Reply:
I wondered that too when writing the guide. It actually does feel like an antipattern, more like a warning when using macros. I wondered if this could be added to the meta-programming guide, maybe after [Write macros responsibly](https://hexdocs.pm/elixir/main/macros.html#write-macros-responsibly) ?

Reply:
In my mind, the core of the issue is that this warning is for application developers first, not library developers. It is hard to think of the best place to put it at right now, besides mix xref. :(

Reply:
Hmm, OK, I get what you're saying. Although I don't agree entirely, I think this is a concern for both application and library developers, albeit a much less important one for libraries.
I'm looking at the `Mix & OTP` section of the documentation. What about linking it indirectly?
1. adding a reference to `mix compile.elixir` in "Introduction to Mix" > "Project compilation" - on a "to go further" kind of link,
2. adding a link to `mix xref` there - there's already a small paragraph about dependencies between files,
3. adding an antipattern section to `mix ref`?

I think it has the advantage of being simple to find. People will search for "compilation" in the Elixir documentation if they're interested in the topic, and they'll find "Project compilation" under "Introduction to Mix". Then they can expand on the topic easily.

Reply:
So to be clear, you want to add more links to the existing `mix xref` docs, right? Would you like to send a PR for those? Also, do you think some of what you wrote could be added to `mix xref`?

Reply:
Indeed, that's the idea : add links to `mix xref` and document this issue in there.
I'll create a new PR for the links and update this one to move this section I wrote in `mix xref` when I get access to my computer again, starting Monday 😄

Reply:
Hi! I'm sorry, I completely forgot about this PR 😅 I updated it to add the links I talked about earlier.